src/Onnx/OnnxToFeld.hs

```diff
@@ -227,51 +246,155 @@ mkInitReadFile hf initGroups = L.concat $ start : map initVec initGroups ++ map
+mkMainFile bname hf weightFile weightRecTC outputs inputs = start
+  where start = L.unlines $
+                [ "#include \"" <> fromString hf <> "\""
+                , ""
+                , "#include <stdlib.h>"
+                , "#include <stdio.h>"
+                , "#include <inttypes.h>"
+                , ""
+                , checkFun
+                , ""
+                , "typedef " <> weightRecTC
+                , "  weight_rec_t;"
+                , ""
+                , "weight_rec_t* read_constants(const char *wfile_name);"
+                , ""
+                , "int main( int argc, char** argv )"
+                , "{"
+                , "  weight_rec_t * w = read_constants( \"" <> fromString weightFile <> "\" );"
+                , ""
+                ]
+                ++ concat (zipWith mkArgRead inputs [1..]) ++
+                [ "  " <> ot <> " " <> ov <> " = {0};"
+                , ""
+                , "  " <> functionName <> "(w, " <> inArgs <> ", &" <> ov <> ");"
+                , ""
+                , oCode
+                , ""
+                , "  return 0;"
+                , "}"
+                ]
+        (ov, ot, oCode) = mkOutput outputs
+        inArgs = L.intercalate ", " ["&" <> mangle (vipName v) | v <- inputs]
+        functionName = fromString $ encodeFunctionName bname
+
+-- | Generate code to read the argument tensors from file
+mkArgRead :: V.ValueInfoProto -> Int -> [L.ByteString]
+mkArgRead vip i = [ "  FILE* " <> fname <> " = fopen( argv[" <> show i <> "], \"r\" );"
+                  , "  " <> tenTypeS <> " " <> vname <> ";"
+                  , ""
+                  , allocReadTensor fname vname n elemT
+                  ]
+  where vname = mangle $ vipName vip
+        fname = "f" <> vname
+        (n, elemT) = vipType vip
+        tenTypeS = fromString $ renderType $ tensorToType n elemT
 
 -- | Compute the (Program) Type that corresponds to a tensor
 tiToType :: TensorInfo -> Type
-tiToType ti = goDim (D.length $ tiDims ti) (toType $ tiType ti)
-  where goDim 0 _  = error "Zero dimensional tensor not implemented"
+tiToType ti = tensorToType (D.length $ tiDims ti) (tiType ti)
+
+-- | Compute the (Program) type corresponding to a tensor with the given dimensionality and element type
+tensorToType :: Int -> TD.DataType -> Type
+tensorToType n t = goDim n $ toType t
+  where goDim 0 _  = error "onnxToFeld.goDim: Zero dimensional tensor not implemented"
         goDim 1 tt = mkAwLType universal tt
-        goDim n tt = mkStructType
+        goDim m tt = mkStructType
                        [("member1", mkAwLType universal tt),
-                        ("member2", mkStructType $ map dimField [1 .. n])
+                        ("member2", mkStructType $ map dimField [1 .. m])
                        ]
-        dimField n = ("member" ++ show n, 1 :# NumType Unsigned S32)
+        dimField i = ("member" ++ show i, 1 :# NumType Unsigned S32)
```

Comment:
I don't like that we have multiple places in the code computing the names of struct members, but I can't find an accessor function for it right now. We do have `uint32` for the type name in Imperative.Frontend though.

Reply:
Same as you, I did not like it but saw no easy way around. But uint32 at least makes it a little better.